File: Data Structures & Algorithms/count-number-of-islands/submission-3.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        ROWS = len(grid)
        COLS = len(grid[0])
        directions = [[1, 0], [0, 1], [0, -1], [-1, 0]]
        visited = set()
        def dfs(r, c):
            if (r < 0 or r >= ROWS or c < 0 or c>= COLS or grid[r][c] == '0' or (r, c) in visited):
                return 

            visited.add((r, c))
            for dr, dc in directions:
                nr, nc = r + dr, c + dc
                dfs(nr, nc)

        res = 0
        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == "1" and (r, c) not in visited:
                    dfs(r, c)
                    res += 1

        return res

File: Data Structures & Algorithms/longest-increasing-path-in-matrix/submission-3.cpp
class Solution {
    int maxres = 0;
    vector<vector<int>> dp;
    vector<vector<int>> directions = {{-1, 0}, {1, 0},
                                      {0, -1}, {0, 1}};

public:
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        int rows = matrix.size();
        int cols = matrix[0].size();
        dp = std::vector<std::vector<int>>(rows, vector<int>(cols, - 1));
        int maxres = 0;
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                maxres = max(dfs(matrix, r, c, -1), maxres);
            }
        }
        return maxres;
    }
private:
    int dfs(vector<vector<int>>& matrix, int r, int c, int prevalue) {
        int rows = matrix.size();
        int cols = matrix[0].size();
        if (r < 0 || c < 0 || r >= rows || c >= cols || matrix[r][c] <= prevalue) {
            return 0;
        }
        if (dp[r][c] != -1) return dp[r][c];

        int res = 1;
        for (vector<int> d : directions) {
            res = max(res, 1 + dfs(matrix, r + d[0], c + d[1], matrix[r][c]));
        }
        dp[r][c] = res;
        maxres = max(maxres, res);
        return res;
    }
};
